server/bleep/src/webserver/aaa.rs

```diff
@@ -0,0 +1,263 @@
+type State = String;
+
+/// Initiate a new login using a web-based OAuth flow.
+///
+/// Redirects to Github after setting up the correct expectations.
+/// Use this endpoint as a redirect target instead of an API call!
+#[utoipa::path(get, path = "/auth/login/start",
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn login(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+) -> impl IntoResponse {
+    let state = rand::thread_rng()
+        .sample_iter(Alphanumeric)
+        .take(STATE_LEN)
+        .map(|c| c as char)
+        .collect::<String>();
+
+    let client_id = app
+        .config
+        .github_client_id
+        .as_ref()
+        .expect("github client id must be provided")
+        .expose_secret();
+
+    let redirect_uri = format!(
+        "https://{}/auth/login/complete",
+        app.config
+            .instance_domain
+            .as_ref()
+            .expect("instance domain must be provided")
+    );
+
+    let github_oauth_url = &format!(
+        "https://github.com/login/oauth/authorize\
+         ?client_id={client_id}\
+         &state={state}\
+         &allow_signup=false\
+         &redirect_uri={redirect_uri}",
+    );
+
+    auth_layer.initialized_login.insert(state);
+
+    Redirect::to(github_oauth_url)
+}
+
+fn update_cookies(jar: CookieJar, oauth_json: String) -> CookieJar {
+    const ONE_YEAR: Duration = Duration::seconds(60 * 60 * 24 * 365);
+
+    jar.add(
+        Cookie::build(AUTH_TOKEN_COOKIE, oauth_json)
+            .path("/")
+            .max_age(ONE_YEAR)
+            .finish(),
+    )
+    .add(
+        Cookie::build(AUTH_CREATED_AT_COOKIE, unix_time().to_string())
+            .path("/")
+            .max_age(ONE_YEAR)
+            .finish(),
+    )
+}
+
+#[derive(Deserialize)]
+pub(super) struct AuthorizedParams {
+    state: State,
+    code: String,
+}
+
+/// Complete the login flow.
+///
+/// Takes the `state` that has been established previously so we don't leak the actual keys.
+#[utoipa::path(post, path = "/auth/login/complete",
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn authorized(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+    Query(params): Query<AuthorizedParams>,
+    jar: CookieJar,
+) -> impl IntoResponse {
+    let AuthorizedParams { state, code } = params;
+
+    auth_layer
+        .initialized_login
+        .remove(&state)
+        .expect("invalid state key");
```

Comment:
I don't understand why but I always get this error although I checked that the state param is the same

Reply:
Perhaps you called this route multiple times? It should only work once, during a redirect from the GitHub login page. Future calls will fail as the key is removed.

Reply:
could be